### DISCUSSION TITLE
Export Component type to user

### DISCUSSION THREAD

**Discussion by liruifengv** [withastro/starlight#2010] [Feature Requests]
### What version of `starlight` are you using?

0.24.2

### What is your idea?

Export Component type to user.

```astro
// user-components/Aside.astro
export interface Props {
	type?: (typeof asideVariants)[number];
	title?: string;
}
```

```ts
// components.ts
export { default as Aside, type Props as AsideProps } from './user-components/Aside.astro';
```

User can import type from `@astrojs/starlight/components`

```astro
---
import { type AsideProps } from "@astrojs/starlight/components"
---
```

### Why is this feature necessary?

When the user wants to encapsulate Starlight component, must use `Parameters`:

```astro
---
import { Badge as StarlightBadge } from '@astrojs/starlight/components';
export type Props = Parameters<typeof StarlightBadge>[0];
---

<StarlightBadge {...Astro.props} />

<style is:global>
	.neutral-badge {
		border-color: var(--sl-color-gray-5);
		background-color: var(--sl-color-gray-6);
		color: var(--sl-color-text);
	}
</style>

```

### Do you have examples of this feature in other projects?

```astro
---
import { Badge as StarlightBadge, type BadgeProps } from '@astrojs/starlight/components';
export type Props = BadgeProps;
---

<StarlightBadge {...Astro.props} />

<style is:global>
	.neutral-badge {
		border-color: var(--sl-color-gray-5);
		background-color: var(--sl-color-gray-6);
		color: var(--sl-color-text);
	}
</style>

```

### Participation

- [ ] I am willing to submit a pull request for this feature.

**Comment by HiDeoo** [comment]
> When the user wants to encapsulate Starlight component, must use `Parameters`:

Another alternative is to use the [`ComponentProps`](https://docs.astro.build/en/guides/typescript/#componentprops-type) type from Astro:

```tsx
import type { ComponentProps } from 'astro/types';
import { Badge } from '@astrojs/starlight/components';
type Prop = ComponentProps<typeof Badge>;
```

**Reply by delucis** [reply]
Yes, given we have the barrel file, I suspect this might be a nicer way to do this. Otherwise we have all the different named components also exporting props. What do folks think? We could also document this better, maybe as part of https://github.com/withastro/starlight/discussions/1938

**Reply by HiDeoo** [reply]
> We could also document this better, maybe as part of #1938

That would make sense in the general "Using components" guide that would be part of this 👍

**Comment by HiDeoo** [comment]
The [`ComponentProps`](https://docs.astro.build/en/guides/typescript/#componentprops-type) approach is now [documented](https://starlight.astro.build/components/using-components/#component-props) in the "Using components" guide.
